packages/compass-aggregations/src/modules/pipeline.ts

```diff
@@ -726,30 +758,124 @@ export const generatePipelineAsString = (state: RootState, index: number) => {
-  const stage = state.pipeline[index];
+const getCancelableExecuteAggDispatch = <
+  D extends ThunkDispatch<RootState, unknown, AnyAction> = ThunkDispatch<
+    RootState,
+    unknown,
+    AnyAction
+  >
+>(
+  id: string,
+  _dispatch: D
+): D => {
+  let canceled = false;
+  ExecuteAggInflightMap.get(id)?.();
+  ExecuteAggInflightMap.set(id, () => {
+    canceled = true;
+  });
+  const dispatch = (action: any) => {
+    if (canceled) return;
+    _dispatch(action);
+  };
+  return dispatch as D;
+};
+
+const ExecuteAggInflightMap = new Map<string, () => void>();
+
+const ExecuteAggDebounceMap = new Map<
+  string,
+  typeof _executeAggregation & { cancel(): void }
+>();
+
+const _executeAggregation = (
+  dataService: DataService,
+  ns: string,
+  dispatch: Dispatch,
+  getState: () => RootState,
+  index: number
+) => {
+  const stage = getState().pipeline[index];
   stage.executor = generateStage(stage);
   if (
     stage.isValid &&
     stage.isEnabled &&
     stage.stageOperator &&
-    (stage.stageOperator !== OUT && stage.stageOperator !== MERGE)
+    stage.stageOperator !== OUT &&
+    stage.stageOperator !== MERGE
   ) {
-    executeStage(dataService, ns, dispatch, state, index);
+    executeStage(dataService, ns, dispatch, getState, index);
   } else {
-    dispatch(stagePreviewUpdated([], index, null, false, state.env));
+    dispatch(stagePreviewUpdated([], index, null, false, getState().env));
+  }
+};
+
+const getDebouncedExecuteAgg = (id: string): typeof _executeAggregation => {
+  if (ExecuteAggDebounceMap.has(id)) {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    return ExecuteAggDebounceMap.get(id)!;
+  } else {
+    const fn = debounce(_executeAggregation, 750, {
+      leading: true,
+      trailing: true
+    });
+    ExecuteAggDebounceMap.set(id, fn);
+    return fn;
   }
 };
 
+/**
+ * Execute the aggregation pipeline at the provided index. The previous execute
+ * request will be canceled in cases when another is dispatched while the
+ * previous one is in-flight. We are not really canceling an operation that is
+ * running on the server here, just making sure that when it resolves the state
+ * update is ignored to prevent race condition state mismatches
+ *
+ * @param {DataService} dataService - The data service.
+ * @param {String} ns - The namespace.
+ * @param {Function} dispatch - The dispatch function.
+ * @param {Object} state - The state.
+ * @param {Number} index - The current index.
+ */
+const executeAggregation = (
+  dataService: DataService,
+  ns: string,
+  _dispatch: Dispatch,
+  getState: () => RootState,
+  index: number
+) => {
+  const id = `${getState().instanceId}::${index}`;
+  const dispatch = getCancelableExecuteAggDispatch(id, _dispatch);
+  const debouncedExecuteAgg = getDebouncedExecuteAgg(id);
+  debouncedExecuteAgg(dataService, ns, dispatch, getState, index);
+};
```

Comment:
So the issue caught by e2e tests in CI was caused by the fact that compass-aggregation plugin is just all full of async flow race-condition related issues that are partially hidden from the end users by very arbitrary `shouldComponentUpdate` checks and extremely weird side effects in `componentDidUpdate` in plugin components and moving code around a bit yet again surfaced some of those problems like it happened a few times already.

There is no way of fixing all of these without changing literally everything about how the plugin state managed and how components are rendered, so for now this seems to be the minimal amount of changes that should make sure that there is at least some normalized async flow for the cases where previews are enabled and stages are changed.